crates/ingress-kafka/src/consumer_task.rs

```diff
@@ -457,36 +305,161 @@ impl Drop for AbortOnDrop {
-            Ok(msg) => msg,
-            Err(err) => break err.into(),
-        };
-        let offset = msg.offset();
-        if let Err(err) = sender.send(&consumer_group_id, msg).await {
-            break err;
+}
+
+impl<T, C> TopicConsumptionTask<T, C>
+where
+    T: TransportConnect,
+    C: ConsumerContext,
+{
+    fn new(
+        ingress: IngressClient<T>,
+        builder: EnvelopeBuilder,
+        topic_partition: TopicPartition,
+        topic_partition_consumer: StreamPartitionQueue<C>,
+        consumer: Arc<MessageConsumer<T>>,
+        consumer_group_id: String,
+        failed: mpsc::UnboundedSender<Error>,
+    ) -> Self {
+        Self {
+            ingress,
+            builder,
+            topic_partition,
+            topic_partition_consumer,
+            consumer,
+            consumer_group_id,
+            failed,
+        }
+    }
+
+    async fn run(mut self) {
+        // this future will be aborted when the partition is no longer needed, so any exit is a failure
+        if let Err(err) = self.inner().await {
+            _ = self.failed.send(err);
         }
-        if let Err(err) = consumer.store_offset(&topic_partition.0, topic_partition.1, offset) {
-            break err.into();
+    }
+
+    async fn inner(&mut self) -> Result<(), Error> {
+        debug!(
+            restate.subscription.id = %self.builder.subscription().id(),
+            messaging.consumer.group.name = self.consumer_group_id,
+            "Starting topic '{}' partition '{}' consumption loop",
+            self.topic_partition.0,
+            self.topic_partition.1
+        );
+
+        let producer_id = dedup_producer_id(
+            &self.consumer_group_id,
+            &self.topic_partition.0,
+            self.topic_partition.1,
+        );
+
+        let ingress_request_counter = counter!(
+            KAFKA_INGRESS_REQUESTS,
+            "subscription" => self.builder.subscription().id().to_string(),
+            "topic" => self.topic_partition.0.to_string(),
+            "partition" => self.topic_partition.1.to_string(),
+        );
+
+        let mut inflight = VecDeque::new();
+
+        loop {
+            tokio::select! {
+                biased;
+                Some(committed) = Self::head_committed(&mut inflight) => {
+                    let head = inflight.pop_front().expect("to exist");
+                    if let Err(CommitError::Cancelled) = committed {
+                        return Err(Error::IngressClosed);
+                    }
+                    ingress_request_counter.increment(1);
+                    trace!(
+                        topic=%self.topic_partition.0, kafka_partition=%self.topic_partition.1, offset=%head.offset,
+                        "store kafka offset",
+                    );
+                    self.consumer.store_offset(&self.topic_partition.0, self.topic_partition.1, head.offset)?;
+                },
+                received = self.topic_partition_consumer.recv() => {
```

Comment:
Slightly unrelated: Wondering whether turning the `topic_partition_consumer` into a stream via `stream` and the consuming from this stream is slightly more efficient. This could also allow to bulk read the ready messages.